incubator/hnc/pkg/forest/forest.go

```diff
@@ -0,0 +1,169 @@
+// Package forest defines the Forest type.
+package forest
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// Forest defines a forest of namespaces - that is, a set of trees. It includes methods to mutate
+// the forest legally (ie, prevent cycles).
+//
+// The forest should always be locked/unlocked (via the `Lock` and `Unlock` methods) while it's
+// being mutated to avoid different controllers from making inconsistent changes.
+type Forest struct {
+	lock       sync.Mutex
+	namespaces namedNamespaces
+}
+
+func NewForest() *Forest {
+	return &Forest{
+		lock:       sync.Mutex{},
+		namespaces: namedNamespaces{},
+	}
+}
+
+func (f *Forest) Lock() {
+	f.lock.Lock()
+}
+
+func (f *Forest) Unlock() {
+	f.lock.Unlock()
+}
+
+// AddOrGet returns a `Namespace` object representing an _existing_ namespace in K8s. Since it
+// already exists, that means that it must be legal according to K8s (ie be unique, have a
+// reasonable name, etc), so this method does very little checking.
+func (f *Forest) AddOrGet(nm string) *Namespace {
+	if nm == "" {
+		// Basically impossible, K8s doesn't allow unnamed objects
+		return nil
+	}
+	ns, ok := f.namespaces[nm]
+	if ok {
+		return ns
+	}
+	ns = &Namespace{forest: f, name: nm, children: namedNamespaces{}}
+	f.namespaces[nm] = ns
+	return ns
+}
+
+// Get returns a `Namespace` object if it exists, or nil if it's missing.
+func (f *Forest) Get(nm string) *Namespace {
+	ns, _ := f.namespaces[nm]
+	return ns
+}
+
+// Remove deletes the given namespace and returns a list of all namespaces who
+// have been affected (ie, the parent and the now-orphaned children).
+func (f *Forest) Remove(nm string) []string {
+	affected := []string{}
+	ns, _ := f.namespaces[nm]
+	if ns == nil {
+		return nil
+	}
+
+	// Update the parent, if any
+	p := ns.Parent()
+	if p != nil {
+		affected = append(affected, p.name)
+		delete(p.children, nm)
+	}
+
+	// Unset the parent from all children, if any
+	for _, c := range ns.children {
+		affected = append(affected, c.name)
+		c.parent = nil
+	}
+
+	// Delete and return affected
+	delete(f.namespaces, nm)
```

Comment:
If this fails here we are in a bad state because all children have been removed but the parent would still remain. The failure is also not caught/handled so a reconcile wont know to try again later.

Reply:
While I'm certain there are many problems with deletion, I don't really understand this one? The in-memory forest is only supposed to _reflect_ what's on the cluster - it's not the source of truth. On the cluster, all the "child" namespaces will continue to have pointers to the now-missing parent, so if we set the resync period to be something reasonable, we'll eventually find the problem.

I'm also not certain if there's a _specific_ failure that you had in mind here - is there?

Reply:
Sorry, in my rush to review this before the weekend I missed that this was the built in and thought this called the controllers delete.